### PR TITLE
Add RemoveEmptyDelete rule

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

**Comment by findepi** [comment]
Fixes #7173

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EmptyDelete.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+/**
+ * If the predicate for a delete is optimized to false, the target table scan
+ * of the delete will be replaced with an empty values node. This type of
+ * plan cannot be executed and is meaningless anyway, so we replace the
+ * entire thing with a values node.
+ */
+public class EmptyDelete
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EmptyDelete.java]: `RemoveEmptyDelete`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EmptyDelete.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        if (!(exchangeSource instanceof DeleteNode)) {
+            return Optional.empty();
+        }
+        DeleteNode delete = (DeleteNode) exchangeSource;
+
+        PlanNode deleteSource = lookup.resolve(delete.getSource());
+        if (!(deleteSource instanceof ValuesNode)) {
+            return Optional.empty();
+        }
+        ValuesNode values = (ValuesNode) deleteSource;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EmptyDelete.java]: I do not like this cascade of `if`s.

What if we write several rules instead? One to replace `DeleteNode` with empty `ValuesNode` to empty `ValuesNode`. Exchange with empty sources to empty node and etc. That way we could rewrite (optimize) more plans - more generic way.

**Comment by kokosing** [comment]
Add tests!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EmptyDelete.java (hunk 2 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EmptyDelete.java]: I can't split this into multiple rules. Replacing (exchange with zero values) with just (zero values) yields `Cannot overwrite partitioning with SINGLE (currently set to Optional[HASH])`. I'm going to stay with the rule as is now.

**Comment by findepi** [comment]
I've added some tests, but needed to rebase onto @losipiuk's #7256, so please review starting at "Add method for Delete-related nodes to PlanMatchPattern" (so basically first commit of mine).

I wanted to add a better test, in `TestLogicalPlanner`, but that class is tied to TCP-H connector, which does not support DELETE. @losipiuk suggested to add a test-dedicated connector that will pretend to have DELETE support. Presto tests cannot use other connectors, since connectors' tests depend on main tests...

@martint, what do you think about such "mocky" connector for tests only?

**Comment by kokosing** [comment]
What about using black hole connector? Please, don't give up so fast. I see
no reason why splitting this rule into several smaller ones should not
work. I think you had  something wrong, will help you tomorrow.

czw., 2.02.2017, 18:09 użytkownik Piotr Findeisen <notifications@github.com>
napisał:

**Comment by findepi** [comment]
The blackhole connector does not support deletes either, so I can't have planning-testing tests using that.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ * plan cannot be executed and is meaningless anyway, so we replace the
+ * entire thing with a values node.
+ */
+public class RemoveEmptyDelete
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+    {
+        if (!(node instanceof TableFinishNode)) {
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java]: If you're into some 'nasty, functional, no more ifs' code, have a look at:
https://github.com/Teradata/presto/commit/250c0155323a247af6edddbac9115910d8765526
https://github.com/Teradata/presto/commit/bec2b021507a593664f980f208284f3599459286
https://github.com/Teradata/presto/commit/9062e93aaed39e2f2c8971e813435c4247a093ff

also https://github.com/Teradata/presto/commits/feature/rule_migrations

cc @kokosing

**Comment by kokosing** [comment]
So why not to add support for DELETE to black hole?. Instead of creating a
new connector.

czw., 2.02.2017, 20:49 użytkownik Piotr Findeisen <notifications@github.com>
napisał:

**Comment by losipiuk** [comment]
> So why not to add support for DELETE to black hole?. Instead of creating a
> new connector.

It seemed fishy to me to pretend that functionality is supported by connector which does not support it. Though for blackhole a no-op delete seem like proper implementation. So it might work.

The bigger problem  here is that to test `EmptyDelete` rule you would have to enforce `AddExchanges` to transform:
```
DELETE
   EXCHANGE
      SCAN_FROM_BLACKHOLE
```

to:

```
DELETE
   EXCHANGE
      VALUES()
```

That would happen if there would be no layout returned for scanned table for given constraint. And as for now `BlackHoleMetadata` always return layout for table, it would not work.
To make this approach work it would require test-specific changes to connector, which I do not like.

It seems to me (if we want tests like this) that it would be much nicer to have a mock-connector which you programmatically setup
to perform specific behaviour at the beginning of the test. 

But first question is still valid. Do we want tests like this?

**Comment by kokosing** [comment]
Sounds like an overkill. Although, such programmable connector would be
great for testing other thingsas well.

czw., 2.02.2017, 21:55 użytkownik Łukasz Osipiuk <notifications@github.com>
napisał:

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 4]
```diff
@@ -84,6 +94,48 @@ public FilterNode filter(Expression predicate, PlanNode source)
...
+                        new TestingTableHandle()),
+                symbols,
+                assignments,
+                Optional.empty(),
+                TupleDomain.all(),
+                originalConstraint
+        );
+    }
+
+    public TableFinishNode tableDelete(SchemaTableName schemaTableName, PlanNode deleteSource, Symbol deleteRowId)
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: do you need to pass `deleteRowId`? I think you could omit it in parameter list.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 5]
```diff
@@ -84,6 +94,48 @@ public FilterNode filter(Expression predicate, PlanNode source)
...
+    {
+        TableWriterNode.DeleteHandle deleteHandle = new TableWriterNode.DeleteHandle(
+                new TableHandle(
+                        new ConnectorId("testConnector"),
+                        new TestingTableHandle()),
+                schemaTableName
+        );
+        return new TableFinishNode(
+                idAllocator.getNextId(),
+                exchange(e -> e
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: I do not think it is the way to go.  `TableFinish` may have arbitrary tree underneath (wll not tree as joins are not supported there) with `DeleteNode` in it.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            return Optional.empty();
+        }
+        TableFinishNode finish = (TableFinishNode) node;
+
+        PlanNode source = lookup.resolve(finish.getSource());
+        if (!(source instanceof ExchangeNode)) {
+            return Optional.empty();
+        }
+        ExchangeNode exchange = (ExchangeNode) source;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java]: I think `exchange` existance here is optional.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java (hunk 4 above)
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: I need to match finish node's output symbols with output symbols of produced values, in `TestRemoveEmptyDelete.test`.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java (hunk 5 above)
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: I am not aiming at a generic factory method for table finish nodes here. That's why this method is called `tableDelete`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java (hunk 6 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java]: Perhaps, but let's keep this rule aligned with `EmptyDeleteOptimizer` for now.

**Comment by findepi** [comment]
Comments replied (or applied), @kokosing

**Comment by findepi** [comment]
@martint, I removed all commits I pulled earlier from @losipiuk's #7256 except the one I really need, so this can be reviewed independently now.

**Comment by arhimondr** [comment]
@findepi Could you please apply this [fixup](https://github.com/Teradata/presto/commit/9e9eff7fecfe2a1e9ee20d5ff2cb5774458b2cdc) once https://github.com/prestodb/presto/pull/7413 is merged

**Comment by findepi** [comment]
Superseded by #7530.

> Could you please apply this fixup once #7413 is merged

#7530 is going to be first, so back to you, @arhimondr. The commit doesn't apply now, as {{RuleTester}} is not yet {{Closeable}}.

**Pull request closed without merging**